File: tumbuhin_flutter/lib/shared/services/bill_tracker_service.dart
import '../../../core/api/api_client.dart';
import '../models/bill.dart';

class BillTrackerService {
  final ApiClient _apiClient;
  BillTrackerService(this._apiClient);

  Future<List<Bill>> getBills({Map<String, String>? params}) async {
    final response = await _apiClient.dio.get(
      '/api/v1/bills',
      queryParameters: params,
    );
    return (response.data as List).map((e) => Bill.fromJson(e)).toList();
  }

  Future<Map<String, dynamic>> createBill(Map<String, dynamic> data) async {
    final response = await _apiClient.dio.post('/api/v1/bills', data: data);
    return response.data as Map<String, dynamic>;
  }

  Future<Bill> getBillDetail(String id) async {
    final response = await _apiClient.dio.get('/api/v1/bills/$id');
    return Bill.fromJson(response.data);
  }

  Future<Map<String, dynamic>> updateBill(
    String id,
    Map<String, dynamic> data,
  ) async {
    final response = await _apiClient.dio.patch(
      '/api/v1/bills/$id',
      data: data,
    );
    return response.data as Map<String, dynamic>;
  }

  Future<void> deleteBill(String id) async {
    await _apiClient.dio.delete('/api/v1/bills/$id');
  }

  Future<Map<String, dynamic>> payBill(
    String id,
    Map<String, dynamic> data,
  ) async {
    final response = await _apiClient.dio.post(
      '/api/v1/bills/$id/pay',
      data: data,
    );
    return response.data as Map<String, dynamic>;
  }

  Future<List<BillPayment>> getPayments(String id) async {
    final response = await _apiClient.dio.get('/api/v1/bills/$id/payments');
    return (response.data as List).map((e) => BillPayment.fromJson(e)).toList();
  }

  Future<void> cancelBill(String id) async {
    await _apiClient.dio.patch('/api/v1/bills/$id/cancel');
  }

  Future<BillSummary> getSummary() async {
    final response = await _apiClient.dio.get('/api/v1/bills/summary');
    return BillSummary.fromJson(response.data);
  }
}

File: web/src/lib/api/billTrackerService.ts
import { apiGet, apiPost, apiPatch, apiDelete } from './client';
import type { Bill, BillPayment, BillSummary } from '@/types';

export const billTrackerService = {
  getBills: (params?: Record<string, string>) =>
    apiGet<Bill[]>('/api/v1/bills', params),

  createBill: (data: {
    title: string; amount: number; bill_type: string; due_date: string;
    contact_name?: string; contact_phone?: string; coa_account_id?: string;
    payment_account_id?: string; reminder_days?: number[]; description?: string;
  }) => apiPost<any>('/api/v1/bills', data),

  getBillDetail: (id: string) => apiGet<Bill>(`/api/v1/bills/${id}`),

  updateBill: (id: string, data: Partial<Bill>) =>
    apiPatch<Bill>(`/api/v1/bills/${id}`, data),

  deleteBill: (id: string) => apiDelete<any>(`/api/v1/bills/${id}`),

  payBill: (id: string, data: { amount: number; payment_date?: string; payment_account_id?: string; notes?: string }) =>
    apiPost<any>(`/api/v1/bills/${id}/pay`, data),

  getPayments: (id: string) => apiGet<BillPayment[]>(`/api/v1/bills/${id}/payments`),

  cancelBill: (id: string) => apiPatch<any>(`/api/v1/bills/${id}/cancel`),

  getSummary: () => apiGet<BillSummary>('/api/v1/bills/summary'),
};
